kinesis/kinesis.go

```diff
@@ -29,6 +34,126 @@ const (
 	RecordSizeLimit int = 1 * 1024 * 1024 // 1MB
 )
 
+// NewFastWriter returns an io.Writer that will ideally be able to flush to kinesis
+// as fast as possible. It will:
+//
+// 1. Buffer up to 4 MB if data in memory before flushing to Kinesis.
+// 2. If the buffer is not filled within 1 second, the data in the buffer will
+//    be flushed.
+// 4. If the calls to the Kinesis API become slow and fall behind, Writes will
+//    be dropped in 4mb chunks.
+func NewFastWriter(streamName, partitionKey string) *BufferedWriter {
+	return newFastWriter(NewDefaultWriter(streamName, partitionKey))
+}
+
+func newFastWriter(w io.Writer) *BufferedWriter {
+	// Allow up to 4 Writes to be queued, if the underlying io.Writer falls
+	// behind because of slow Writes, drop it.
+	dw := NewDropWriter(w, 4)
+
+	// Buffer 4mb in memory before flushing to Kinesis and periodically
+	// flush every second.
+	return NewBufferedWriter(dw, 4*1024*1024, time.Second)
+}
+
+type flushWriter interface {
+	io.Writer
+	Flush() error
+}
+
+// BufferedWriter is a wrapper around bufio.Writer that can periodically flush
+// the buffer instead of waiting for it to fill completely.
+type BufferedWriter struct {
+	w  flushWriter
+	mu sync.Mutex // Protects the underlying bufio.Writer, since we'll be flushing in a separate goroutine.
+	t  <-chan time.Time
+}
+
+// NewBufferedDefaultWriter returns a new io.Writer that buffers up to size
+// bytes in memory before flushing to the Kinesis. It will also periodically
+// flush the buffer if it is not full.
+func NewBufferedWriter(w io.Writer, size int, flush time.Duration) *BufferedWriter {
+	wr := &BufferedWriter{
+		w: bufio.NewWriterSize(w, size),
+		t: time.Tick(flush),
+	}
+	go wr.start()
+	return wr
+}
+
+// start receives on the ticker and flushes.
+func (w *BufferedWriter) start() {
+	for range w.t {
+		w.Flush()
```

Comment:
This should really be smarter and only flush if the buffer wasn't flushed within that second.